base/regex.jl

```diff
@@ -174,6 +174,42 @@ function occursin(r::Regex, s::SubString; offset::Integer=0)
                      r.match_data)
 end
 
+"""
+    startswith(s::AbstractString, prefix::Regex)
+
+Return `true` if `s` starts with the Regex pattern, `prefix`.
+
+See also [`endswith`](@ref).
+
+# Examples
+```jldoctest
+julia> startswith("JuliaLang", r"Julia|Romeo")
+true
+```
+"""
+function startswith(s::AbstractString, r::Regex)
+    rr = Regex(r.pattern, r.compile_options | PCRE.ANCHORED, r.match_options)
```

Comment:
This feels a bit like a performance trap?

Typically you write something like `foo(s, r"...")` so that the regex gets created during macro expansion and not over and over during runtime. But this is precisely what this does. Should the `startswith`-regex be cached somehow or is the cost of creating it really insignificant?

Reply:
So, this implementation seems to be around 100 times slower compared to `occursin`, so I think you have a point about it not being "insignificant".

My thoughts were that people would mostly use this function when checking file extensions, etc., but I am all for making it work faster.  Do you have a suggestion for how to implement the cache?  Presumably it should be pretty small? :slightly_smiling_face:

Reply:
For `startswith`, we could pass `PCRE_ANCHORED` to `pcre_exec()`? That way, we would not need to recompile the pattern.

For `endswith`, I don't know. We want to avoid recompiling, but also want to start matching at the end, if at all possible. I don't see an option bit for this.

Reply:
PCRE2 apparently supports anchoring patterns at runtime without recompile; quoting from http://www.pcre.org/pcre2.txt:

>The PCRE2_ANCHORED, PCRE2_ENDANCHORED, and PCRE2_NO_UTF_CHECK options can be set at the time of matching as well as at compile time.

So I think we should use `EXECUTE_MASK` / match-options and add these two options; `occursin` might want a keyword argument for match options, because we don't want to modify the regex (multithreading, avoiding `finally` for restoration after the match in case something went wrong).